experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+#===------------------------------------------------------------------------===#
+# Query types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_ssa.bool
+  '''
+  """
+  name = "rel_ssa.bool"
+
+
+@irdl_attr_definition
+class SchemaElement(ParametrizedAttribute):
+  """
+  Models an element of a schema with name `elt_name` and type `elt_type`.
+
+  Example:
+  '''
+  !rel_ssa.schema_element<"id", !rel_ssa.int32>
+  '''
+  """
+  name = "rel_ssa.schema_element"
+
+  elt_name = ParameterDef(StringAttr)
+  elt_type = ParameterDef(DataType)
+
+  @staticmethod
+  @builder
+  def get(name: str, type_: DataType) -> 'SchemaElement':
+    return SchemaElement([StringAttr.from_str(name), type_])
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
+  """
+  Models a bag in a relational SSA query. The exact schema of the bag is part of
+  the type itself.
+
+  Example:
+
+  '''
+  !rel_ssa.bag<[!rel_ssa.schema_element<"id", !rel_ssa.int32>]>
+  '''
+  """
+  name = "rel_ssa.bag"
+
+  schema = ParameterDef(ArrayOfConstraint(SchemaElement))
+
+  @staticmethod
+  @builder
+  def get(types: list[DataType], names: list[str]) -> 'Bag':
+    schema_elts = [SchemaElement.get(n, t) for n, t in zip(names, types)]
+    return Bag([ArrayAttr.from_list(schema_elts)])
+
+
+#===------------------------------------------------------------------------===#
+# Expressions
+#===------------------------------------------------------------------------===#
+
+
+class Expression(Operation):
+  """
+  Interface class for all Expressions, i.e., operations that work on single
+  values.
+  """
+  ...
+
+
+@irdl_op_definition
+class Column(Expression):
+  """
+  References a specific column with name `col_name` and returns its value. The
+  name 'col_name' must be part of the schema of an encompassing operator.
+
+  Example:
+
+  '''
+  %0 : rel_ssa.int32 = rel_ssa.column() ["col_name" = "id"]
+  '''
+  """
+  name = "rel_ssa.column"
+
+  col_name = AttributeDef(StringAttr)
+  result = ResultDef(DataType)
+
+
+@irdl_op_definition
+class Compare(Expression):
+  """
+  Returns `left` 'comparator' `right`.
+
+  Example:
+
+  '''
+  ...
+  %0 : !rel_ssa.bool = rel_ssa.compare(%a : !rel_ssa.int32, %b : !rel_ssa.int32) ["comparator" = "="]
+  '''
+  """
+  name = "rel_ssa.compare"
+
+  left = OperandDef(DataType)
```

Comment:
Same comment as above, this one really operates on int32 types.
This begs the question of just using `int32` here and even just using a vanilla MLIR func + SymbolAttr that operates on int32.
This op is quite different from the Column op.